File: archive/flutter-app/test/core/utility/jwt_parser_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:infrakit_studio/core/utility/jwt_parser.dart';

/// `{"alg":"HS256","typ":"JWT"}`
const headerHs256 = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9';

/// `{"alg":"none","typ":"JWT"}` — 35 chars, so base64url padding is missing.
const headerNone = 'eyJhbGciOiJub25lIiwidHlwIjoiSldUIn0';

/// The canonical jwt.io demo payload:
/// `{"sub":"1234567890","name":"John Doe","iat":1516239022}`
const payloadClassic = 'eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ';

/// All seven registered claims; `exp` = 1600000000 (2020-09-13T12:26:40Z).
const payloadAllClaims =
    'eyJpc3MiOiJodHRwczovL2F1dGguaW5mcmFraXQudGVzdC8iLCJzdWIiOiJ1c2VyLTQyIiwiYXVkIjoiaW5mcmFraXQt'
    'YXBpIiwiZXhwIjoxNjAwMDAwMDAwLCJuYmYiOjE1MDAwMDAwMDAsImlhdCI6MTUwMDAwMDAwMCwianRpIjoiYWJjLTEyMyJ9';

/// `exp` = 4102444800 (2100-01-01T00:00:00Z).
const payloadFarFuture =
    'eyJpc3MiOiJodHRwczovL2F1dGguaW5mcmFraXQudGVzdC8iLCJzdWIiOiJ1c2VyLTQyIiwiZXhwIjo0MTAyNDQ0ODAwLCJpYXQiOjE1MDAwMDAwMDB9';

/// `{"sub":"early","nbf":4102444800,"exp":4102531200}`
const payloadNotYetValid = 'eyJzdWIiOiJlYXJseSIsIm5iZiI6NDEwMjQ0NDgwMCwiZXhwIjo0MTAyNTMxMjAwfQ';

/// `[1,2,3]` — valid base64url, valid JSON, but not a JSON *object*.
const payloadJsonArray = 'WzEsMiwzXQ';

const signature = 'SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c';

/// A fixed "now" so expiry assertions never depend on the wall clock.
final now = DateTime.utc(2024, 6, 1, 12);

JwtParseResult parse(String token) =>
    const JwtParser().execute(JwtParseInput(token: token, now: now));

void main() {
  group('decoding a known token', () {
    final result = parse('$headerHs256.$payloadClassic.$signature');

    test('reports a structurally valid token', () {
      expect(result.isValidStructure, isTrue);
      expect(result.errorMessage, isNull);
      expect(result.segmentCount, 3);
    });

    test('decodes the header to the expected map', () {
      expect(result.header, {'alg': 'HS256', 'typ': 'JWT'});
      expect(result.algorithm, 'HS256');
      expect(result.tokenType, 'JWT');
      expect(result.keyId, isNull);
    });

    test('decodes the payload to the expected map', () {
      expect(result.payload, {'sub': '1234567890', 'name': 'John Doe', 'iat': 1516239022});
    });

    test('pretty-prints both segments as indented JSON', () {
      expect(result.headerJson, contains('"alg": "HS256"'));
      expect(result.payloadJson, contains('"sub": "1234567890"'));
      expect(result.payloadJson, contains('\n'));
    });

    test('keeps the signature segment intact but does not verify it', () {
      expect(result.signatureBase64Url, signature);
      expect(result.signatureVerified, isFalse);
    });

    test('converts iat to a real DateTime', () {
      expect(result.issuedAt, DateTime.utc(2018, 1, 18, 1, 30, 22));
    });

    test('tolerates a Bearer prefix and wrapped whitespace', () {
      final prefixed = parse('  Bearer $headerHs256.\n$payloadClassic.\n$signature  ');
      expect(prefixed.isValidStructure, isTrue);
      expect(prefixed.payload['sub'], '1234567890');
    });
  });

  group('base64url without padding', () {
    test('decodes a 35-character segment that needs one "=" restored', () {
      // headerNone.length % 4 == 3, so the raw string is not valid base64.
      expect(headerNone.length % 4, 3);
      final result = parse('$headerNone.$payloadClassic.');
      expect(result.isValidStructure, isTrue);
      expect(result.header['alg'], 'none');
    });

    test('decodeBase64Url restores padding directly', () {
      // "eyJhIjoxfQ" is `{"a":1}` with the padding stripped.
      expect(String.fromCharCodes(JwtParser.decodeBase64Url('eyJhIjoxfQ')), '{"a":1}');
    });

    test('decodes base64url -/_ characters, not just +/', () {
      final decoded = JwtParser.decodeBase64Url('-_-_');
      expect(decoded, [251, 255, 191]);
    });
  });

  group('registered claims', () {
    final result = parse('$headerHs256.$payloadAllClaims.$signature');

    test('surfaces all seven in canonical order', () {
      expect(
        result.registeredClaims.map((c) => c.name).toList(),
        ['iss', 'sub', 'aud', 'exp', 'nbf', 'iat', 'jti'],
      );
    });

    test('each carries a human-readable label and meaning', () {
      for (final claim in result.registeredClaims) {
        expect(claim.label, isNotEmpty);
        expect(claim.meaning, isNotEmpty);
      }
      final exp = result.registeredClaims.firstWhere((c) => c.name == 'exp');
      expect(exp.label, 'Expires at');
      expect(exp.meaning, contains('accept'));
    });

    test('NumericDate claims become DateTimes and display as ISO-8601 UTC', () {
      final exp = result.registeredClaims.firstWhere((c) => c.name == 'exp');
      expect(exp.dateTime, DateTime.utc(2020, 9, 13, 12, 26, 40));
      expect(exp.displayValue, '2020-09-13T12:26:40Z UTC');
      expect(exp.rawValue, 1600000000);
    });

    test('non-date claims keep their raw string value', () {
      final iss = result.registeredClaims.firstWhere((c) => c.name == 'iss');
      expect(iss.dateTime, isNull);
      expect(iss.displayValue, 'https://auth.infrakit.test/');
    });

    test('claims that are absent are not invented', () {
      final classic = parse('$headerHs256.$payloadClassic.$signature');
      expect(classic.registeredClaims.map((c) => c.name), ['sub', 'iat']);
    });
  });

  group('expiry detection', () {
    test('an exp in the past is reported as expired with a negative remainder', () {
      final result = parse('$headerHs256.$payloadAllClaims.$signature');
      expect(result.temporalStatus, JwtTemporalStatus.expired);
      expect(result.isExpired, isTrue);
      expect(result.expiresAt, DateTime.utc(2020, 9, 13, 12, 26, 40));
      expect(result.timeUntilExpiry!.isNegative, isTrue);
    });

    test('an exp in the future is still valid with a positive remainder', () {
      final result = parse('$headerHs256.$payloadFarFuture.$signature');
      expect(result.temporalStatus, JwtTemporalStatus.valid);
      expect(result.isExpired, isFalse);
      expect(result.expiresAt, DateTime.utc(2100));
      expect(result.timeUntilExpiry!.isNegative, isFalse);
      expect(result.timeUntilExpiry!.inDays, greaterThan(27000));
    });

    test('an nbf in the future is reported as not-yet-valid', () {
      final result = parse('$headerHs256.$payloadNotYetValid.$signature');
      expect(result.temporalStatus, JwtTemporalStatus.notYetValid);
      expect(result.isNotYetValid, isTrue);
      expect(result.timeUntilValid, isNotNull);
      expect(result.timeUntilValid!.isNegative, isFalse);
    });

    test('a token with no time claims at all is not called expired', () {
      final result = parse('$headerHs256.$payloadClassic.$signature');
      expect(result.temporalStatus, JwtTemporalStatus.valid);
      expect(result.expiresAt, isNull);
      expect(result.timeUntilExpiry, isNull);
    });

    test('exp exactly at "now" counts as expired', () {
      final at = DateTime.utc(2020, 9, 13, 12, 26, 40);
      final result = const JwtParser().execute(
        JwtParseInput(token: '$headerHs256.$payloadAllClaims.$signature', now: at),
      );
      expect(result.temporalStatus, JwtTemporalStatus.expired);
    });

    test('formatJwtDuration renders a readable countdown', () {
      expect(formatJwtDuration(const Duration(seconds: 45)), '45s');
      expect(formatJwtDuration(const Duration(minutes: 90)), '1h 30m');
      expect(formatJwtDuration(const Duration(days: 2, hours: 3)), '2d 3h');
      expect(formatJwtDuration(const Duration(days: -2)), '2d');
    });
  });

  group('alg: none is flagged', () {
    test('an unsecured JWT with an empty signature is flagged', () {
      final result = parse('$headerNone.$payloadClassic.');
      expect(result.isValidStructure, isTrue);
      expect(result.algorithm, 'none');
      expect(result.isUnsignedAlgorithm, isTrue);
    });

    test('alg "NONE" in any casing is flagged', () {
      // {"alg":"None"} base64url-encoded.
      const mixedCase = 'eyJhbGciOiJOb25lIn0';
      final result = parse('$mixedCase.$payloadClassic.');
      expect(result.isUnsignedAlgorithm, isTrue);
    });

    test('a missing third segment counts as unsigned', () {
      final result = parse('$headerHs256.$payloadClassic');
      expect(result.isValidStructure, isTrue);
      expect(result.segmentCount, 2);
      expect(result.isUnsignedAlgorithm, isTrue);
    });

    test('a normal HS256 token is not flagged', () {
      final result = parse('$headerHs256.$payloadClassic.$signature');
      expect(result.isUnsignedAlgorithm, isFalse);
    });
  });

  group('signature is never verified', () {
    test('signatureVerified is false even for a well-formed token', () {
      expect(parse('$headerHs256.$payloadClassic.$signature').signatureVerified, isFalse);
    });

    test('the notice says so plainly', () {
      expect(JwtParseResult.signatureNotice, contains('NOT verified'));
      expect(JwtParseResult.signatureNotice, contains('only decodes'));
      expect(JwtParseResult.signatureNotice.toLowerCase(), contains('forged'));
    });
  });

  group('malformed input errors cleanly, never throws', () {
    void expectsFailure(String token, {String? containing}) {
      final result = parse(token);
      expect(result.isValidStructure, isFalse, reason: 'expected failure for "$token"');
      expect(result.errorMessage, isNotNull);
      expect(result.errorMessage, isNotEmpty);
      if (containing != null) expect(result.errorMessage, contains(containing));
    }

    test('empty input', () => expectsFailure('', containing: 'Enter a JWT'));
    test('whitespace only', () => expectsFailure('   \n  '));

    test('one segment', () => expectsFailure('notajwt', containing: '3 dot-separated'));

    test('four segments', () {
      expectsFailure('$headerHs256.$payloadClassic.$signature.extra', containing: '3 dot-separated');
    });

    test('empty header segment', () => expectsFailure('.$payloadClassic.$signature'));

    test('bad base64 in the header', () {
      expectsFailure('!!!not-base64!!!.$payloadClassic.$signature', containing: 'base64url');
    });

    test('bad base64 in the payload', () {
      expectsFailure('$headerHs256.!!!nope!!!.$signature', containing: 'base64url');
    });

    test('valid base64 that is not JSON', () {
      // "aGVsbG8gd29ybGQ" is "hello world".
      expectsFailure('aGVsbG8gd29ybGQ.$payloadClassic.$signature', containing: 'not valid JSON');
    });

    test('JSON that is not an object', () {
      expectsFailure('$headerHs256.$payloadJsonArray.$signature', containing: 'must be a JSON object');
    });

    test('failures carry no half-populated decoded data', () {
      final result = parse('nope');
      expect(result.header, isEmpty);
      expect(result.payload, isEmpty);
      expect(result.registeredClaims, isEmpty);
      expect(result.headerJson, '');
      expect(result.signatureVerified, isFalse);
    });
  });
}

File: app/src/core/utility/jwtParser.test.ts
import { describe, expect, it } from 'vitest'
import { decodeBase64Url, formatJwtDuration, JWT_SIGNATURE_NOTICE, JwtParser, type JwtParseResult } from './jwtParser'

/** `{"alg":"HS256","typ":"JWT"}` */
const headerHs256 = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9'

/** `{"alg":"none","typ":"JWT"}` — 35 chars, so base64url padding is missing. */
const headerNone = 'eyJhbGciOiJub25lIiwidHlwIjoiSldUIn0'

/**
 * The canonical jwt.io demo payload:
 * `{"sub":"1234567890","name":"John Doe","iat":1516239022}`
 */
const payloadClassic = 'eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ'

/** All seven registered claims; `exp` = 1600000000 (2020-09-13T12:26:40Z). */
const payloadAllClaims =
  'eyJpc3MiOiJodHRwczovL2F1dGguaW5mcmFraXQudGVzdC8iLCJzdWIiOiJ1c2VyLTQyIiwiYXVkIjoiaW5mcmFraXQt' +
  'YXBpIiwiZXhwIjoxNjAwMDAwMDAwLCJuYmYiOjE1MDAwMDAwMDAsImlhdCI6MTUwMDAwMDAwMCwianRpIjoiYWJjLTEyMyJ9'

/** `exp` = 4102444800 (2100-01-01T00:00:00Z). */
const payloadFarFuture =
  'eyJpc3MiOiJodHRwczovL2F1dGguaW5mcmFraXQudGVzdC8iLCJzdWIiOiJ1c2VyLTQyIiwiZXhwIjo0MTAyNDQ0ODAwLCJpYXQiOjE1MDAwMDAwMDB9'

/** `{"sub":"early","nbf":4102444800,"exp":4102531200}` */
const payloadNotYetValid = 'eyJzdWIiOiJlYXJseSIsIm5iZiI6NDEwMjQ0NDgwMCwiZXhwIjo0MTAyNTMxMjAwfQ'

/** `[1,2,3]` — valid base64url, valid JSON, but not a JSON *object*. */
const payloadJsonArray = 'WzEsMiwzXQ'

const signature = 'SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c'

/** A fixed "now" so expiry assertions never depend on the wall clock. */
const now = new Date(Date.UTC(2024, 5, 1, 12))

function parse(token: string): JwtParseResult {
  return new JwtParser().execute({ token, now })
}

describe('JwtParser', () => {
  describe('decoding a known token', () => {
    const result = parse(`${headerHs256}.${payloadClassic}.${signature}`)

    it('reports a structurally valid token', () => {
      expect(result.isValidStructure).toBe(true)
      expect(result.errorMessage).toBeUndefined()
      expect(result.segmentCount).toBe(3)
    })

    it('decodes the header to the expected map', () => {
      expect(result.header).toEqual({ alg: 'HS256', typ: 'JWT' })
      expect(result.algorithm).toBe('HS256')
      expect(result.tokenType).toBe('JWT')
      expect(result.keyId).toBeUndefined()
    })

    it('decodes the payload to the expected map', () => {
      expect(result.payload).toEqual({ sub: '1234567890', name: 'John Doe', iat: 1516239022 })
    })

    it('pretty-prints both segments as indented JSON', () => {
      expect(result.headerJson).toContain('"alg": "HS256"')
      expect(result.payloadJson).toContain('"sub": "1234567890"')
      expect(result.payloadJson).toContain('\n')
    })

    it('keeps the signature segment intact but does not verify it', () => {
      expect(result.signatureBase64Url).toBe(signature)
    })

    it('converts iat to a real Date', () => {
      expect(result.issuedAt).toEqual(new Date(Date.UTC(2018, 0, 18, 1, 30, 22)))
    })

    it('tolerates a Bearer prefix and wrapped whitespace', () => {
      const prefixed = parse(`  Bearer ${headerHs256}.\n${payloadClassic}.\n${signature}  `)
      expect(prefixed.isValidStructure).toBe(true)
      expect(prefixed.payload['sub']).toBe('1234567890')
    })
  })

  describe('base64url without padding', () => {
    it('decodes a 35-character segment that needs one "=" restored', () => {
      // headerNone.length % 4 == 3, so the raw string is not valid base64.
      expect(headerNone.length % 4).toBe(3)
      const result = parse(`${headerNone}.${payloadClassic}.`)
      expect(result.isValidStructure).toBe(true)
      expect(result.header['alg']).toBe('none')
    })

    it('decodeBase64Url restores padding directly', () => {
      // "eyJhIjoxfQ" is `{"a":1}` with the padding stripped.
      const bytes = decodeBase64Url('eyJhIjoxfQ')
      expect(new TextDecoder().decode(bytes)).toBe('{"a":1}')
    })

    it('decodes base64url -/_ characters, not just +/', () => {
      const decoded = decodeBase64Url('-_-_')
      expect(Array.from(decoded)).toEqual([251, 255, 191])
    })
  })

  describe('registered claims', () => {
    const result = parse(`${headerHs256}.${payloadAllClaims}.${signature}`)

    it('surfaces all seven in canonical order', () => {
      expect(result.registeredClaims.map((c) => c.name)).toEqual(['iss', 'sub', 'aud', 'exp', 'nbf', 'iat', 'jti'])
    })

    it('each carries a human-readable label and meaning', () => {
      for (const claim of result.registeredClaims) {
        expect(claim.label.length).toBeGreaterThan(0)
        expect(claim.meaning.length).toBeGreaterThan(0)
      }
      const exp = result.registeredClaims.find((c) => c.name === 'exp')!
      expect(exp.label).toBe('Expires at')
      expect(exp.meaning).toContain('accept')
    })

    it('NumericDate claims become Dates and display as ISO-8601 UTC', () => {
      const exp = result.registeredClaims.find((c) => c.name === 'exp')!
      expect(exp.dateTime).toEqual(new Date(Date.UTC(2020, 8, 13, 12, 26, 40)))
      expect(exp.displayValue).toBe('2020-09-13T12:26:40Z UTC')
      expect(exp.rawValue).toBe(1600000000)
    })

    it('non-date claims keep their raw string value', () => {
      const iss = result.registeredClaims.find((c) => c.name === 'iss')!
      expect(iss.dateTime).toBeUndefined()
      expect(iss.displayValue).toBe('https://auth.infrakit.test/')
    })

    it('claims that are absent are not invented', () => {
      const classic = parse(`${headerHs256}.${payloadClassic}.${signature}`)
      expect(classic.registeredClaims.map((c) => c.name)).toEqual(['sub', 'iat'])
    })
  })

  describe('expiry detection', () => {
    it('an exp in the past is reported as expired with a negative remainder', () => {
      const result = parse(`${headerHs256}.${payloadAllClaims}.${signature}`)
      expect(result.temporalStatus).toBe('expired')
      expect(result.expiresAt).toEqual(new Date(Date.UTC(2020, 8, 13, 12, 26, 40)))
      expect(result.timeUntilExpiryMs!).toBeLessThan(0)
    })

    it('an exp in the future is still valid with a positive remainder', () => {
      const result = parse(`${headerHs256}.${payloadFarFuture}.${signature}`)
      expect(result.temporalStatus).toBe('valid')
      expect(result.expiresAt).toEqual(new Date(Date.UTC(2100, 0, 1)))
      expect(result.timeUntilExpiryMs!).toBeGreaterThanOrEqual(0)
      expect(result.timeUntilExpiryMs! / 86400000).toBeGreaterThan(27000)
    })

    it('an nbf in the future is reported as not-yet-valid', () => {
      const result = parse(`${headerHs256}.${payloadNotYetValid}.${signature}`)
      expect(result.temporalStatus).toBe('notYetValid')
      expect(result.timeUntilValidMs).not.toBeUndefined()
      expect(result.timeUntilValidMs!).toBeGreaterThanOrEqual(0)
    })

    it('a token with no time claims at all is not called expired', () => {
      const result = parse(`${headerHs256}.${payloadClassic}.${signature}`)
      expect(result.temporalStatus).toBe('valid')
      expect(result.expiresAt).toBeUndefined()
      expect(result.timeUntilExpiryMs).toBeUndefined()
    })

    it('exp exactly at "now" counts as expired', () => {
      const at = new Date(Date.UTC(2020, 8, 13, 12, 26, 40))
      const result = new JwtParser().execute({ token: `${headerHs256}.${payloadAllClaims}.${signature}`, now: at })
      expect(result.temporalStatus).toBe('expired')
    })

    it('formatJwtDuration renders a readable countdown', () => {
      expect(formatJwtDuration(45 * 1000)).toBe('45s')
      expect(formatJwtDuration(90 * 60 * 1000)).toBe('1h 30m')
      expect(formatJwtDuration((2 * 24 + 3) * 3600 * 1000)).toBe('2d 3h')
      expect(formatJwtDuration(-2 * 24 * 3600 * 1000)).toBe('2d')
    })
  })

  describe('alg: none is flagged', () => {
    it('an unsecured JWT with an empty signature is flagged', () => {
      const result = parse(`${headerNone}.${payloadClassic}.`)
      expect(result.isValidStructure).toBe(true)
      expect(result.algorithm).toBe('none')
      expect(result.isUnsignedAlgorithm).toBe(true)
    })

    it('alg "NONE" in any casing is flagged', () => {
      // {"alg":"None"} base64url-encoded.
      const mixedCase = 'eyJhbGciOiJOb25lIn0'
      const result = parse(`${mixedCase}.${payloadClassic}.`)
      expect(result.isUnsignedAlgorithm).toBe(true)
    })

    it('a missing third segment counts as unsigned', () => {
      const result = parse(`${headerHs256}.${payloadClassic}`)
      expect(result.isValidStructure).toBe(true)
      expect(result.segmentCount).toBe(2)
      expect(result.isUnsignedAlgorithm).toBe(true)
    })

    it('a normal HS256 token is not flagged', () => {
      const result = parse(`${headerHs256}.${payloadClassic}.${signature}`)
      expect(result.isUnsignedAlgorithm).toBe(false)
    })
  })

  describe('signature is never verified', () => {
    it('the notice says so plainly', () => {
      expect(JWT_SIGNATURE_NOTICE).toContain('NOT verified')
      expect(JWT_SIGNATURE_NOTICE).toContain('only decodes')
      expect(JWT_SIGNATURE_NOTICE.toLowerCase()).toContain('forged')
    })
  })

  describe('malformed input errors cleanly, never throws', () => {
    function expectsFailure(token: string, containing?: string) {
      const result = parse(token)
      expect(result.isValidStructure).toBe(false)
      expect(result.errorMessage).not.toBeUndefined()
      expect(result.errorMessage!.length).toBeGreaterThan(0)
      if (containing !== undefined) expect(result.errorMessage).toContain(containing)
    }

    it('empty input', () => expectsFailure('', 'Enter a JWT'))
    it('whitespace only', () => expectsFailure('   \n  '))

    it('one segment', () => expectsFailure('notajwt', '3 dot-separated'))

    it('four segments', () => {
      expectsFailure(`${headerHs256}.${payloadClassic}.${signature}.extra`, '3 dot-separated')
    })

    it('empty header segment', () => expectsFailure(`.${payloadClassic}.${signature}`))

    it('bad base64 in the header', () => {
      expectsFailure(`!!!not-base64!!!.${payloadClassic}.${signature}`, 'base64url')
    })

    it('bad base64 in the payload', () => {
      expectsFailure(`${headerHs256}.!!!nope!!!.${signature}`, 'base64url')
    })

    it('valid base64 that is not JSON', () => {
      // "aGVsbG8gd29ybGQ" is "hello world".
      expectsFailure(`aGVsbG8gd29ybGQ.${payloadClassic}.${signature}`, 'not valid JSON')
    })

    it('JSON that is not an object', () => {
      expectsFailure(`${headerHs256}.${payloadJsonArray}.${signature}`, 'must be a JSON object')
    })

    it('failures carry no half-populated decoded data', () => {
      const result = parse('nope')
      expect(result.header).toEqual({})
      expect(result.payload).toEqual({})
      expect(result.registeredClaims).toEqual([])
      expect(result.headerJson).toBe('')
    })
  })
})
